src/librustc/sir.rs

```diff
@@ -0,0 +1,136 @@
+// (SirFuncIdx, SirBlockIdx) pair must be used.
+newtype_index! {
+    pub struct SirBlockIdx {
+        DEBUG_FORMAT = "SirBlockIdx({})"
+    }
+}
+
+/// Sir equivalents of LLVM values.
+#[derive(Debug)]
+pub enum SirValue {
+    Func(SirFuncIdx),
+}
+
+impl SirValue {
+    pub fn func_idx(&self) -> SirFuncIdx {
+        let Self::Func(idx) = self;
+        *idx
+    }
+}
+
+pub struct SirCx {
+    /// Maps an opaque LLVM `Value` to its SIR equivalent.
+    pub llvm_values: FxHashMap<*const Value, SirValue>,
+    /// Maps an opaque LLVM `BasicBlock` to the function and block index of it's SIR equivalent.
+    pub llvm_blocks: FxHashMap<*const BasicBlock, (SirFuncIdx, SirBlockIdx)>,
+    /// Function store. Also owns the blocks
+    pub funcs: IndexVec<SirFuncIdx, ykpack::Body>,
+}
+
+impl SirCx {
+    pub fn new() -> Self {
+        Self {
+            llvm_values: Default::default(),
+            llvm_blocks: FxHashMap::default(),
+            funcs: Default::default(),
+        }
+    }
+
+    pub fn add_func(&mut self, value: *const Value, symbol_name: String) {
+        let idx = SirFuncIdx::from_usize(self.funcs.len());
+
+        self.funcs.push(ykpack::Body{
+            symbol_name,
+            blocks: Default::default(),
+            num_locals: 0,  // FIXME
+            num_args: 0,    // FIXME
+            flags: 0,       // Set later.
+        });
+        let existing = self.llvm_values.insert(value, SirValue::Func(idx));
+        debug_assert!(existing.is_none());
+    }
+
+    pub fn add_block(&mut self, func: *const Value, block: *const BasicBlock) {
+        let func_idx = self.llvm_values[&func].func_idx();
+        let sir_func = &mut self.funcs[func_idx];
+        let block_idx = SirBlockIdx::from_usize(sir_func.blocks.len());
+        sir_func.blocks.push(ykpack::BasicBlock{
+            stmts: Default::default(),
+            term: ykpack::Terminator::Unreachable, // FIXME
+        });
+        let existing = self.llvm_blocks.insert(block, (func_idx, block_idx));
+        debug_assert!(existing.is_none());
+    }
+
+    pub fn get_symbol_name(&mut self, func: *const Value) -> &String {
+        let func_idx = self.llvm_values[&func].func_idx();
+        let sir_func = &mut self.funcs[func_idx];
+        &sir_func.symbol_name
+    }
+
+    /// For hardware tracing, during codegen we insert DILabels to know where we are in the binary.
+    /// These labels must be emitted in a deterministic order otherwise the reproducible build
+    /// checker gets upset. This function gives the codegen what it needs in a data structure which
+    /// can be iterated deterministically.
+    pub fn funcs_and_blocks_deterministic(&self)
+        -> IndexVec<SirFuncIdx, IndexVec<SirBlockIdx, *const BasicBlock>>
+    {
+        // Scaffold a data structure which can be iterated deterministically by the caller.
+        // To start, all pointer addresses are None. We would have use overwritable null pointers,
```

Comment:
"we would have TO use"?

Reply:
I also don't understand the `None` / "overwriteable null pointers" comparison.

Reply:
The "scaffolding" here builds a data structure with mutable place-holder `None`s which are updated to `Some`s in the loop later. Rather than using `Option<*const ...>` and stripping the resulting `Some`s later, I'd have liked to have used `null` as the placeholder, but Rust refuses.

Reply:
OK. Can we maybe just tweak the comment a bit then?